habitat_baselines/common/rollout_storage.py

```diff
@@ -4,237 +4,196 @@
+        )
         if action_space.__class__.__name__ == "ActionSpace":
             action_shape = 1
         else:
             action_shape = action_space.shape[0]
 
-        self.actions = torch.zeros(num_steps, num_envs, action_shape)
-        self.prev_actions = torch.zeros(num_steps + 1, num_envs, action_shape)
+        self.buffers["actions"] = torch.zeros(
+            numsteps + 1, num_envs, action_shape
+        )
+        self.buffers["prev_actions"] = torch.zeros(
+            numsteps + 1, num_envs, action_shape
+        )
         if action_space.__class__.__name__ == "ActionSpace":
-            self.actions = self.actions.long()
-            self.prev_actions = self.prev_actions.long()
+            self.buffers["actions"] = self.buffers["actions"].long()
+            self.buffers["prev_actions"] = self.buffers["prev_actions"].long()
+
+        self.buffers["masks"] = torch.zeros(
+            numsteps + 1, num_envs, 1, dtype=torch.bool
+        )
+
+        self.is_double_buffered = is_double_buffered
+        self._nbuffers = 2 if is_double_buffered else 1
+        self._num_envs = num_envs
+
+        assert (self._num_envs % self._nbuffers) == 0
 
-        self.masks = torch.zeros(num_steps + 1, num_envs, 1)
+        self.numsteps = numsteps
+        self.steps = [0 for _ in range(self._nbuffers)]
 
-        self.num_steps = num_steps
-        self.step = 0
+    @property
+    def step(self):
+        assert all(s == self.steps[0] for s in self.steps)
+        return self.steps[0]
 
     def to(self, device):
-        for sensor in self.observations:
-            self.observations[sensor] = self.observations[sensor].to(device)
-
-        self.recurrent_hidden_states = self.recurrent_hidden_states.to(device)
-        self.rewards = self.rewards.to(device)
-        self.value_preds = self.value_preds.to(device)
-        self.returns = self.returns.to(device)
-        self.action_log_probs = self.action_log_probs.to(device)
-        self.actions = self.actions.to(device)
-        self.prev_actions = self.prev_actions.to(device)
-        self.masks = self.masks.to(device)
+        self.buffers.map_in_place(lambda v: v.to(device))
 
     def insert(
         self,
-        observations,
-        recurrent_hidden_states,
-        actions,
-        action_log_probs,
-        value_preds,
-        rewards,
-        masks,
+        observations=None,
+        recurrent_hidden_states=None,
+        actions=None,
+        action_log_probs=None,
+        value_preds=None,
+        rewards=None,
+        masks=None,
+        buffer_index: int = 0,
     ):
-        for sensor in observations:
-            self.observations[sensor][self.step + 1].copy_(
-                observations[sensor]
-            )
-        self.recurrent_hidden_states[self.step + 1].copy_(
-            recurrent_hidden_states
+        if not self.is_double_buffered:
+            assert buffer_index == 0
+
+        next_step = dict(
```

Comment:
Maybe, that worth to divide next vs current step parameters using prefix as well `next_observations`, ....

Reply:
Makes sense